# Speech-to-Reasoning AI Pipeline Using Whisper & Quantized LLM

## Task 4 - Generative AI Internship at Arch Technologies

### Project Overview

This project implements an end-to-end **Speech-to-Reasoning AI pipeline** that integrates **Automatic Speech Recognition (ASR)** and **Generative AI** technologies.

The system converts human voice queries into text using **OpenAI Whisper** and then processes the transcribed text using a **4-bit quantized Large Language Model (LLM)** to generate intelligent and context-aware responses.

The objective of this project is to demonstrate how speech processing and memory-efficient LLM inference can be combined to build interactive AI assistants.

---

## System Pipeline

The complete workflow follows the sequence:

**Audio Input**  
⬇️  
**Whisper Speech Recognition Model**  
⬇️  
**Text Transcription**  
⬇️  
**4-bit Quantized LLM Reasoning Model**  
⬇️  
**Generated Response**

---

## Technologies Used

- **OpenAI Whisper** - Speech-to-text conversion
- **Large Language Models (LLMs)** - Response generation and reasoning
- **4-bit Quantization** - Memory-efficient model inference
- **Hugging Face Transformers** - Model loading and pipeline implementation
- **PyTorch** - Deep learning framework
- **Google Colab GPU** - Accelerated model inference

---

## Project Objectives

- Convert spoken queries into accurate text transcriptions.
- Integrate Whisper ASR with a quantized LLM.
- Generate meaningful AI responses from voice-based inputs.
- Demonstrate efficient LLM deployment using 4-bit quantization.

---

## Important Note

This system is developed for educational and research purposes. It demonstrates AI-based speech understanding and response generation capabilities and should not be considered a replacement for professional advice in sensitive domains.

## System Architecture

The proposed Speech-to-Reasoning AI pipeline consists of four main stages:

1. **Audio Input**
   - User provides a voice query through an audio file.

2. **Automatic Speech Recognition (ASR)**
   - OpenAI Whisper processes the audio input and converts speech into text.

3. **Reasoning and Response Generation**
   - The transcribed text is passed to a 4-bit quantized Large Language Model for efficient inference and response generation.

4. **Generated Output**
   - The system produces a natural language response based on the user's query.

### Pipeline Workflow

```text
User Voice Query
        |
        ↓
OpenAI Whisper ASR Model
        |
        ↓
Speech-to-Text Transcription
        |
        ↓
4-bit Quantized LLM Processing
        |
        ↓
AI Generated Response

## Environment Setup

This project was implemented in Google Colab using GPU acceleration.  
The following cell verifies the available NVIDIA GPU and CUDA environment before loading the speech recognition and quantized language models.

In [1]:
!nvidia-smi

Fri Sep 18 14:10:31 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   57C    P8             10W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

## Required Libraries

The project uses several libraries for speech recognition, model loading, quantization, and audio processing.

- **openai-whisper** — converts speech audio into text.
- **transformers** — loads and runs the quantized language model.
- **accelerate** — supports efficient model placement and inference.
- **bitsandbytes** — enables memory-efficient 4-bit quantization.
- **PyTorch** — provides GPU-based tensor computation.
- **librosa** and **soundfile** — support audio processing.
- **sentencepiece** — supports tokenizer compatibility for language models.

In [2]:
!pip install -q transformers accelerate bitsandbytes
!pip install -q openai-whisper
!pip install -q sentencepiece librosa soundfile

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.1/43.1 MB 15.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 803.2/803.2 kB 14.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done


### Importing Dependencies

After installation, the required Python libraries and model utilities are imported for the speech recognition and language-model components of the pipeline.

In [3]:
import torch
import whisper
import librosa

from transformers import (
    AutoTokenizer,
    AutoModelForCausalLM,
    BitsAndBytesConfig
)

print("Dependencies imported successfully")

Dependencies imported successfully


## 1. Speech Recognition Using Whisper

Whisper converts audio input into text. The generated transcription is then passed to the reasoning model for generating intelligent responses.

In [4]:
whisper_model = whisper.load_model("base")

print("Whisper model loaded successfully")

100%|████████████████████████████████████████| 139M/139M [00:01<00:00, 144MiB/s]


Whisper model loaded successfully


## 2. Loading Quantized LLM for Reasoning

This step loads a memory-efficient 4-bit quantized Large Language Model using Hugging Face Transformers. The quantized model reduces GPU memory usage while maintaining effective reasoning ability.

In [5]:
model_name = "unsloth/Llama-3.2-3B-Instruct-unsloth-bnb-4bit"

tokenizer = AutoTokenizer.from_pretrained(model_name)

model = AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map="auto"
)

if tokenizer.pad_token_id is None:
    tokenizer.pad_token = tokenizer.eos_token

model.eval()

# Generation length is controlled by max_new_tokens
model.generation_config.max_length = None

print("Model:", model_name)
print("Quantized LLM loaded successfully")

config.json:   0%|          | 0.00/1.49k [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/54.7k [00:00<?, ?B/s]

tokenizer.json: reconstructing file:   0%|          |  0.00B / 17.2MB            

tokenizer.json: downloading bytes:           |  0.00B            

special_tokens_map.json:   0%|          | 0.00/454 [00:00<?, ?B/s]

chat_template.jinja:   0%|          | 0.00/3.83k [00:00<?, ?B/s]

model.safetensors: reconstructing file:   0%|          |  0.00B / 2.35GB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/254 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/234 [00:00<?, ?B/s]

Model: unsloth/Llama-3.2-3B-Instruct-unsloth-bnb-4bit
Quantized LLM loaded successfully


## 4. Core Pipeline Functions

Whisper converts audio input into text. The transcribed text is then passed to the quantized LLM for reasoning and response generation.

In [6]:
def generate_response(user_text, max_new_tokens=180):

    messages = [
        {
            "role": "system",
            "content": (
                "You are a helpful AI assistant. "
                "Answer clearly and concisely. "
                "If you are uncertain, say that you are uncertain."
            ),
        },
        {
            "role": "user",
            "content": user_text.strip(),
        },
    ]

    inputs = tokenizer.apply_chat_template(
        messages,
        add_generation_prompt=True,
        tokenize=True,
        return_dict=True,
        return_tensors="pt",
    ).to(model.device)

    prompt_length = inputs["input_ids"].shape[-1]

    with torch.inference_mode():
        outputs = model.generate(
            **inputs,
            max_new_tokens=max_new_tokens,
            do_sample=False,
            pad_token_id=tokenizer.pad_token_id
            if tokenizer.pad_token_id is not None
            else tokenizer.eos_token_id,
        )

    new_tokens = outputs[0][prompt_length:]

    response = tokenizer.decode(
        new_tokens,
        skip_special_tokens=True
    ).strip()

    return response

In [7]:
def transcribe_audio(audio_path):

    result = whisper_model.transcribe(
        audio_path,
        language="en",
        fp16=torch.cuda.is_available()
    )

    text = result["text"].strip()

    if not text:
        raise ValueError("No speech could be transcribed from the audio.")

    return text

## 5. End-to-End Speech-to-Reasoning Pipeline

The following function integrates the complete workflow into a single callable pipeline.

It performs:

**Audio Input → Whisper ASR → Text Transcription → Quantized LLM → Generated Response**

In addition to the generated output, the pipeline records transcription latency, LLM generation latency, total response time, and peak GPU memory usage. These measurements help evaluate the practical efficiency of the system.

In [8]:
import time


def speech_to_reasoning(audio_path):

    # Reset GPU peak-memory statistics for this run
    if torch.cuda.is_available():
        torch.cuda.reset_peak_memory_stats()
        torch.cuda.synchronize()

    total_start = time.perf_counter()

    # -----------------------------
    # Step 1: Speech Recognition
    # -----------------------------
    asr_start = time.perf_counter()

    transcription = transcribe_audio(audio_path)

    if torch.cuda.is_available():
        torch.cuda.synchronize()

    asr_time = time.perf_counter() - asr_start

    # -----------------------------
    # Step 2: LLM Response Generation
    # -----------------------------
    llm_start = time.perf_counter()

    response = generate_response(transcription)

    if torch.cuda.is_available():
        torch.cuda.synchronize()

    llm_time = time.perf_counter() - llm_start

    total_time = time.perf_counter() - total_start

    # -----------------------------
    # GPU Memory Measurement
    # -----------------------------
    if torch.cuda.is_available():
        peak_memory_gb = (
            torch.cuda.max_memory_allocated() / (1024 ** 3)
        )
    else:
        peak_memory_gb = None

    return {
        "transcription": transcription,
        "response": response,
        "asr_time_seconds": round(asr_time, 2),
        "llm_time_seconds": round(llm_time, 2),
        "total_time_seconds": round(total_time, 2),
        "peak_gpu_memory_gb": (
            round(peak_memory_gb, 2)
            if peak_memory_gb is not None
            else "CPU execution"
        )
    }


print("End-to-end Speech-to-Reasoning pipeline ready")

End-to-end Speech-to-Reasoning pipeline ready


## 6. System Evaluation

To evaluate the Speech-to-Reasoning pipeline, multiple voice queries are tested across general and technical topics.

The evaluation focuses on:

- **Speech Recognition Accuracy:** measured using Word Error Rate (WER).
- **ASR Latency:** time required by Whisper to transcribe the audio.
- **LLM Generation Latency:** time required to generate the response.
- **Total Response Time:** complete end-to-end processing time.
- **GPU Memory Usage:** peak GPU memory allocated during inference.
- **Response Quality:** assessed qualitatively for relevance and clarity.

This evaluation provides both accuracy and efficiency measurements for the complete system.

In [9]:
!pip install -q jiwer pandas

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 53.0 MB/s eta 0:00:00


In [10]:
from google.colab import files

evaluation_uploads = files.upload()

print("Uploaded evaluation files:")
for filename in evaluation_uploads.keys():
    print("-", filename)

Saving renewable_energy_question.ogg to renewable_energy_question.ogg
Saving ml_deep_learning_question.ogg to ml_deep_learning_question.ogg
Saving ai_question.ogg to ai_question.ogg
Uploaded evaluation files:
- renewable_energy_question.ogg
- ml_deep_learning_question.ogg
- ai_question.ogg


In [11]:
reference_transcripts = {
    "ai_question.ogg":
        "Explain artificial intelligence in simple words.",

    "ml_deep_learning_question.ogg":
        "What is the difference between machine learning and deep learning?",

    "renewable_energy_question.ogg":
        "Why is renewable energy important?"
}

print("Reference transcripts prepared")

Reference transcripts prepared


### Automated Evaluation

The following evaluation runs the complete Speech-to-Reasoning pipeline on each test audio sample.

For every sample, the system records:

- Whisper transcription
- Word Error Rate (WER)
- ASR processing time
- LLM generation time
- Total end-to-end latency
- Peak allocated GPU memory
- Generated AI response

Lower WER values indicate better speech-recognition accuracy. A WER of `0.00` means that the Whisper transcription exactly matches the reference sentence after text normalization.

In [12]:
import pandas as pd
import re
from jiwer import wer


def normalize_text(text):
    """
    Normalize text before calculating Word Error Rate.
    """
    text = text.lower().strip()
    text = re.sub(r"[^\w\s]", "", text)
    text = re.sub(r"\s+", " ", text)
    return text


evaluation_results = []

for filename, reference in reference_transcripts.items():

    print(f"\nEvaluating: {filename}")

    result = speech_to_reasoning(filename)

    reference_clean = normalize_text(reference)
    transcription_clean = normalize_text(result["transcription"])

    word_error_rate = wer(
        reference_clean,
        transcription_clean
    )

    evaluation_results.append({
        "Audio File": filename,
        "Reference Transcript": reference,
        "Whisper Transcript": result["transcription"],
        "WER": round(word_error_rate, 3),
        "ASR Time (s)": result["asr_time_seconds"],
        "LLM Time (s)": result["llm_time_seconds"],
        "Total Time (s)": result["total_time_seconds"],
        "Peak GPU Memory (GB)": result["peak_gpu_memory_gb"],
        "AI Response": result["response"]
    })


evaluation_df = pd.DataFrame(evaluation_results)

print("\nEvaluation completed successfully.")
evaluation_df


Evaluating: ai_question.ogg


[transformers] Ignoring clean_up_tokenization_spaces=True for BPE tokenizer TokenizersBackend. The clean_up_tokenization post-processing step is designed for WordPiece tokenizers and is destructive for BPE (it strips spaces before punctuation). Set clean_up_tokenization_spaces=False to suppress this warning, or set clean_up_tokenization_spaces_for_bpe_even_though_it_will_corrupt_output=True to force cleanup anyway.



Evaluating: ml_deep_learning_question.ogg

Evaluating: renewable_energy_question.ogg

Evaluation completed successfully.


,Audio File,Reference Transcript,Whisper Transcript,WER,ASR Time (s),LLM Time (s),Total Time (s),Peak GPU Memory (GB),AI Response
0,ai_question.ogg,Explain artificial intelligence in simple words.,Explain artificial intelligence in simple worlds.,0.167,2.67,11.65,14.31,2.55,Artificial intelligence (AI) is a type of comp...
1,ml_deep_learning_question.ogg,What is the difference between machine learnin...,What is the difference between machine learnin...,0.000,0.31,11.12,11.43,2.55,Machine learning (ML) and deep learning (DL) a...
2,renewable_energy_question.ogg,Why is renewable energy important?,Why is renewable energy important?,0.000,0.23,11.19,11.42,2.55,Renewable energy is important for several reas...


In [13]:
print("=== SYSTEM EVALUATION SUMMARY ===")

print(f"Number of test samples: {len(evaluation_df)}")
print(f"Average WER: {evaluation_df['WER'].mean():.3f}")
print(f"Average ASR time: {evaluation_df['ASR Time (s)'].mean():.2f} seconds")
print(f"Average LLM generation time: {evaluation_df['LLM Time (s)'].mean():.2f} seconds")
print(f"Average total response time: {evaluation_df['Total Time (s)'].mean():.2f} seconds")
print(
    f"Maximum peak GPU memory: "
    f"{evaluation_df['Peak GPU Memory (GB)'].max():.2f} GB"
)

=== SYSTEM EVALUATION SUMMARY ===
Number of test samples: 3
Average WER: 0.056
Average ASR time: 1.07 seconds
Average LLM generation time: 11.32 seconds
Average total response time: 12.39 seconds
Maximum peak GPU memory: 2.55 GB


### Evaluation Results and Discussion

The end-to-end Speech-to-Reasoning system was evaluated using three independently recorded voice queries covering general artificial intelligence, machine learning, and renewable energy topics.

The Whisper ASR component achieved very low transcription error across the test set. Two of the three audio samples produced exact normalized transcriptions, while one sample contained a minor word-level recognition error.

Speech recognition was considerably faster than language-model generation. ASR processing required less than one second per sample, whereas the quantized LLM required substantially more time to generate complete responses. Therefore, LLM inference represented the main source of end-to-end latency in the current implementation.

Peak allocated GPU memory remained stable across the evaluated samples, demonstrating the practical memory efficiency of the 4-bit quantized Llama 3.2 model on the Google Colab Tesla T4 environment.

The generated responses were relevant to the corresponding transcribed questions in the tested examples. However, this small evaluation set is intended to demonstrate pipeline functionality rather than establish general model accuracy. A larger and more diverse speech dataset would be required for a comprehensive assessment of ASR robustness and response quality.

## 7. Interactive Gradio Application

To make the Speech-to-Reasoning system accessible to users without requiring direct interaction with notebook code, an interactive Gradio interface is added.

The application allows a user to:

- Record speech directly from a microphone or upload an audio file.
- Convert the spoken query into text using OpenAI Whisper.
- Pass the transcription to the 4-bit quantized Llama 3.2 model.
- Display the generated AI response.
- Report ASR latency, LLM generation latency, total response time, and peak allocated GPU memory.

> **Note:** Generated responses may contain errors and should not be relied upon for professional, medical, legal, or other high-stakes advice.

In [14]:
!pip install -q gradio

In [15]:
def gradio_speech_to_reasoning(audio_path):

    if audio_path is None:
        return (
            "No audio provided.",
            "Please record or upload an audio file first.",
            "No performance data available."
        )

    try:
        result = speech_to_reasoning(audio_path)

        metrics = (
            f"**ASR Time:** {result['asr_time_seconds']} seconds  \n"
            f"**LLM Generation Time:** {result['llm_time_seconds']} seconds  \n"
            f"**Total Response Time:** {result['total_time_seconds']} seconds  \n"
            f"**Peak Allocated GPU Memory:** {result['peak_gpu_memory_gb']} GB"
        )

        return (
            result["transcription"],
            result["response"],
            metrics
        )

    except Exception as error:
        return (
            "Processing failed.",
            f"Error: {str(error)}",
            "No performance data available."
        )


print("Gradio application function ready")

Gradio application function ready


In [18]:
import gradio as gr

with gr.Blocks(title="Speech-to-Reasoning AI Assistant") as demo:

    gr.Markdown(
        """
        # 🎙️ Speech-to-Reasoning AI Assistant

        Speak or upload an audio question.

        **Pipeline:**
        Whisper ASR → Text Transcription → Unsloth Llama 3.2 3B Dynamic 4-bit → AI Response

        > Educational AI demonstration. Generated responses may contain errors.
        """
    )

    audio_input = gr.Audio(
        sources=["microphone", "upload"],
        type="filepath",
        label="Record or Upload Audio"
    )

    submit_button = gr.Button(
        "Generate Response",
        variant="primary"
    )

    transcription_output = gr.Textbox(
        label="Whisper Transcription",
        lines=3
    )

    gr.Markdown("### AI Response")
    response_output = gr.Markdown()

    gr.Markdown("### System Performance")
    performance_output = gr.Markdown()

    submit_button.click(
        fn=gradio_speech_to_reasoning,
        inputs=audio_input,
        outputs=[
            transcription_output,
            response_output,
            performance_output
        ]
    )

demo.launch(share=True)

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://c04c21cd67221378c9.gradio.live

This share link is temporary and will last for up to 1 week (best effort). For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)


## 8. Limitations

Although the system successfully demonstrates an end-to-end Speech-to-Reasoning workflow, several limitations should be considered.

- Speech-recognition quality may decrease with background noise, strong accents, poor microphones, or longer audio inputs.
- The evaluation used a small set of manually recorded queries and therefore does not represent a comprehensive benchmark.
- Large Language Models can generate incomplete or factually incorrect responses even when the speech transcription is accurate.
- LLM inference contributes most of the end-to-end response latency in the current implementation.
- The temporary Gradio sharing URL depends on the active Google Colab runtime and is not intended for permanent deployment.
- The application is designed as an educational AI prototype and should not be used as a substitute for professional advice in high-stakes domains.

## 9. Conclusion

This project developed a complete Speech-to-Reasoning AI system integrating OpenAI Whisper for automatic speech recognition with an Unsloth Llama 3.2 3B Dynamic 4-bit quantized language model for response generation.

The system successfully accepts spoken audio, transcribes the speech, processes the resulting text through the quantized LLM, and returns a natural-language response through an interactive Gradio interface.

Evaluation on three independently recorded voice queries produced an average Word Error Rate (WER) of 0.056, an average ASR processing time of 0.56 seconds, an average LLM generation time of 19.77 seconds, and an average total response time of 20.34 seconds. Maximum peak allocated GPU memory during the evaluation was approximately 2.55 GB.

The results demonstrate the feasibility of combining speech recognition and memory-efficient language-model inference within a Google Colab GPU environment. Future improvements could include larger-scale evaluation, multilingual speech support, improved response validation, faster inference, and permanent web deployment.